cms/djangoapps/contentstore/views/certificates.py

```diff
@@ -0,0 +1,380 @@
+
+
+class CertificateManager(object):
+    """
+    The CertificateManager is responsible for storage, retrieval, and manipulation of Certificates
+    Certificates are not stored in the Django ORM, they are a field/setting on the course descriptor
+    """
+    @staticmethod
+    def parse(json_string):
+        """
+        Deserialize the provided JSON data into a standard Python object
+        """
+        try:
+            certificate = json.loads(json_string)
+        except ValueError:
+            raise CertificateValidationError(_("invalid JSON"))
+        # Include the data contract version
+        certificate["version"] = CERTIFICATE_SCHEMA_VERSION
+        # Ensure a signatories list is always returned
+        if certificate.get("signatories") is None:
+            certificate["signatories"] = []
+        return certificate
+
+    @staticmethod
+    def validate(certificate_data):
+        """
+        Ensure the certificate data contains all of the necessary fields and the values match our rules
+        """
+        # Ensure the schema version meets our expectations
+        if certificate_data.get("version") != CERTIFICATE_SCHEMA_VERSION:
+            raise TypeError(
+                "Unsupported certificate schema version: {0}.  Expected version: {1}.".format(
+                    certificate_data.get("version"),
+                    CERTIFICATE_SCHEMA_VERSION
+                )
+            )
+        if not certificate_data.get("name"):
+            raise CertificateValidationError(_("must have name of the certificate"))
+
+    @staticmethod
+    def get_used_ids(course):
+        """
+        Return a list of certificate identifiers that are already in use for this course
+        """
+        if not course.certificates or not course.certificates.get('certificates'):
+            return []
+        return [cert['id'] for cert in course.certificates['certificates']]
+
+    @staticmethod
+    def assign_id(course, certificate_data, certificate_id=None):
+        """
+        Assign an identifier to the provided certificate data.
+        If the caller did not provide an identifier, we autogenerate a unique one for them
+        In addition, we check the certificate's signatories and ensure they also have unique ids
+        """
+        used_ids = CertificateManager.get_used_ids(course)
+        if certificate_id:
+            certificate_data['id'] = int(certificate_id)
+        else:
+            certificate_data['id'] = generate_int_id(
+                CERTIFICATE_MINIMUM_ID,
+                MYSQL_MAX_INT,
+                used_ids
+            )
+
+        for index, signatory in enumerate(certificate_data['signatories']):  # pylint: disable=unused-variable
+            if signatory and not signatory.get('id', False):
+                signatory['id'] = generate_int_id(used_ids=used_ids)
+            used_ids.append(signatory['id'])
+
+        return certificate_data
+
+    @staticmethod
+    def to_json(certificate):
```

Comment:
Nit / Optional: `to_json` and `from_json` both seem a little oddly named because strictly speaking they are `to_dict` and `from_dict`